### PR TITLE
feat(http-proxy-middleware): remove types

### LINKED ISSUE #45479: [webpack-plugin-serve] http-proxy-middleware ships with types that are favored over @types/http-proxy-middleware and are not the same
- [x] I tried using the `@types/xxxx` package and had problems.
- [x] I tried using the latest stable version of tsc. https://www.npmjs.com/package/typescript
- [x] I have a question that is inappropriate for [StackOverflow](https://stackoverflow.com/).  (Please ask any appropriate questions there).
- [x] [Mention](https://github.com/blog/821-mention-somebody-they-re-notified) the authors (see `Definitions by:` in `index.d.ts`) so they can respond.
  - Authors: @shellscape @PlayMa256 @chimurai @zebMcCorkle @BendingBender

I've pinged all the parties involved, as this is an interesting situation that I think might require some coordination.

`http-proxy-middleware` now ships with its own types, which don't export the same types are `@types/http-proxy-middleware` which are imported by `@types/webpack-plugin-serve`.

This causes TypeScript to error since it favors types from the original package over one's sourced from `@types`:

```
node_modules/@types/webpack-plugin-serve/index.d.ts:9:10 - error TS2305: Module '"../../http-proxy-middleware/dist"' has no exported member 'Config'.

9 import { Config as HttpProxyMiddlewareConfig, Proxy } from 'http-proxy-middleware';
           ~~~~~~

node_modules/@types/webpack-plugin-serve/index.d.ts:9:47 - error TS2305: Module '"../../http-proxy-middleware/dist"' has no exported member 'Proxy'.

9 import { Config as HttpProxyMiddlewareConfig, Proxy } from 'http-proxy-middleware';
                                                ~~~~~
```

For the first error it seems that the problem is mainly one of naming: the `Config` types equivalent in `http-proxy-middleware` is called `Options`, so that should be fixable by having `@types/http-proxy-middleware` re-export the `Config` type as `Options`, and then import the type under that name in `@types/webpack-plugin-serve`.

However looking further into this for the `Proxy` type, it seems that `Proxy` is equal to `ReturnType<createProxyMiddleware>` exported from `http-proxy-middleware`, and so actually the whole import can be replaced with just importing `createProxyMiddleware` and the `proxy` property type becoming `typeof createProxyMiddleware`....

As can probably be seen this is a bit of an interesting situation that I don't know the best way to solve without breaking at least something for some version combo.

I think a good path forward could be to replace the dependency on `@types/http-proxy-middleware` with a peerDependency on `http-proxy-middleware@1` and import in `createProxyMiddleware`, but am open to other suggestions.

-------

This can be reproduced by doing `npm i -D @types/webpack-plugin-serve webpack-plugin-serve typescript && touch index.ts && tsc --noEmit index.ts`

### PR DESCRIPTION
**Pull request by G-Rath**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If removing a declaration:
- [x] If a package was never on Definitely Typed, you don't need to do anything. (If you wrote a package and provided types, you don't need to register it with us.)
- [x] Delete the package's directory.
- [x] Add it to `notNeededPackages.json`.

-----

#51103 is failing because CI keeps grabbing these types, so let's just remove them 🎉 

Fixes #45479 
Closes #51103

### CODE REVIEW

**Review by BendingBender — APPROVED** [review]

**Comment by elibarzilay** [comment]
@G-Rath, should this be merged or wait for someone to look it over?

File: types/webpack-dev-server/package.json [hunk 1]
```diff
@@ -0,0 +1,6 @@
+{
+  "private": true,
+  "dependencies": {
+    "http-proxy-middleware": "^1.0.0"
```
Comment (peterblazejewicz) [inline types/webpack-dev-server/package.json:4]: IMO this cannot be done without package upgrade itself, the version 3.1 does not support that version of
`http-proxy-middleware`:
https://github.com/webpack/webpack-dev-server/blob/v3.11.2/package.json#L48

**Comment by G-Rath** [comment]
@elibarzilay this shoud be good to go!

**Comment by elibarzilay** [comment]
@G-Rath Despite the version difference that @peterblazejewicz pointed at?

**Comment by G-Rath** [comment]
@elibarzilay yes, I think that's fine:

  * this is the only way to unbreak `webpack-plugin-serve` types
  * `webpack-dev-server` is using v1 of `http-proxy-middleware` in their upcoming v4
  * the types don't appear to have any major differences so they should still work
    * otherwise, the types are small enough that we could easily inline them into `webpack-dev-server` if it caused trouble

**Comment by razorness** [comment]
This merge broke projects with dependency of `webpack-dev-server` `<4.0.0` and probably many other projects which are depending on this typings.

```
ERROR in /app/node_modules/@types/webpack-dev-server/index.d.ts(17,38):
TS7016: Could not find a declaration file for module 'http-proxy-middleware'. 
/app/node_modules/http-proxy-middleware/index.js' implicitly has an 'any' type.
```

A quick fix is to create your own typings with this content:

```typescript
declare module 'http-proxy-middleware' {

	import * as http from 'http';
	import * as net from 'net';
	import * as tls from 'tls';
	import * as connect from 'connect';
	import * as httpProxy from 'http-proxy';
	import { Readable } from 'stream';

	const foo: any;
	export default foo;
	export type Filter = (pathname: string, req: http.IncomingMessage) => boolean;
	export type Logger = (...args: any[]) => void;

	export interface LogProvider {
		log: Logger;
		debug?: Logger;
		info?: Logger;
		warn?: Logger;
		error?: Logger;
	}

	export interface Config {
		pathRewrite?: { [ regexp: string ]: string } | ((path: string, req: http.IncomingMessage) => string);
		router?: { [ hostOrPath: string ]: string } | ((req: http.IncomingMessage) => string);
		logLevel?: 'debug' | 'info' | 'warn' | 'error' | 'silent';

		logProvider?(provider: LogProvider): LogProvider;

		onError?(err: Error, req: http.IncomingMessage, res: http.ServerResponse): void;

		onProxyRes?(proxyRes: http.IncomingMessage, req: http.IncomingMessage, res: http.ServerResponse): void;

		onProxyReq?(proxyReq: http.ClientRequest, req: http.IncomingMessage, res: http.ServerResponse): void;

		onProxyReqWs?(proxyReq: http.ClientRequest, req: http.IncomingMessage, socket: net.Socket, options: httpProxy.ServerOptions, head: any): void;

		onOpen?(proxySocket: net.Socket): void;

		onClose?(res: http.IncomingMessage, socket: net.Socket, head: any): void;

		target?: string;
		forward?: string;
		agent?: http.Agent;
		ssl?: tls.TlsOptions;
		ws?: boolean;
		xfwd?: boolean;
		secure?: boolean;
		toProxy?: boolean;
		prependPath?: boolean;
		ignorePath?: boolean;
		localAddress?: string;
		changeOrigin?: boolean;
		preserveHeaderKeyCase?: boolean;
		auth?: string;
		hostRewrite?: string;
		autoRewrite?: boolean;
		protocolRewrite?: string;
		cookieDomainRewrite?: false | string | { [ domain: string ]: string };
		cookiePathRewrite?: false | string | { [ path: string ]: string };
		headers?: { [ header: string ]: string };
		proxyTimeout?: number;
		timeout?: number;
		followRedirects?: boolean;
		selfHandleResponse?: boolean;
		buffer?: Readable;
		/**
		 * @deprecated
		 */
		proxyHost?: any;
		/**
		 * @deprecated
		 */
		proxyTable?: any;
	}

	export type Proxy = connect.NextHandleFunction;
}
```

**Comment by G-Rath** [comment]
@razorness I'm not sure how you're getting that error since this changes `@types/webpack-dev-server` to pull in `http-proxy-middleware` as a dependency to ensure its types are available, and previous versions have a dependency on `@types/http-proxy-middleware` so I'm not really sure how you could be getting that error unless your dependency tree is in an invalid state (i.e it's only half installed).

Regardless, you should be able to resolve this error by explicitly pulling in `@types/http-proxy-middleware` :)

**Comment by razorness** [comment]
As far as I can see `@types/http-proxy-middleware` doesn't exist anymore and yarn was not able to pull it.

https://www.npmjs.com/search?q=%40types%2Fhttp-proxy-middleware&csrftoken=rLHoppKmU9oHCADYZHGlUgjfZIWVNj19bhOnHqoPM10

**Comment by G-Rath** [comment]
[It very much does still exist](https://www.npmjs.com/package/@types/http-proxy-middleware) :)

**Comment by razorness** [comment]
Thank you.
Okay... when I delete lock file and pull this explicitly to my project by `yarn add --dev @types/http-proxy-middleware`, this error gets thrown:

```
ERROR in /app/node_modules/@types/webpack-dev-server/index.d.ts(30,53):
TS2694: Namespace '"/app/node_modules/http-proxy-middleware/dist/index"' has no exported member 'Config'.
ERROR in /app/node_modules/@types/webpack-dev-server/index.d.ts(36,29):
TS2694: Namespace '"/app/node_modules/http-proxy-middleware/dist/index"' has no exported member 'Config'.
Version: typescript 4.2.2, eslint 7.22.0
```

yarn.lock looks like this:
```
"@types/webpack-dev-server@3.11.0":
  version "3.11.0"
  resolved "https://registry.yarnpkg.com/@types/webpack-dev-server/-/webpack-dev-server-3.11.0.tgz#bcc3b85e7dc6ac2db25330610513f2228c2fcfb2"
  integrity sha512-3+86AgSzl18n5P1iUP9/lz3G3GMztCp+wxdDvVuNhx1sr1jE79GpYfKHL8k+Vht3N74K2n98CuAEw4YPJCYtDA==
  dependencies:
    "@types/connect-history-api-fallback" "*"
    "@types/express" "*"
    "@types/http-proxy-middleware" "*"
    "@types/serve-static" "*"
    "@types/webpack" "*"
// ...
"@types/http-proxy-middleware@*", "@types/http-proxy-middleware@^1.0.0":
  version "1.0.0"
  resolved "https://registry.yarnpkg.com/@types/http-proxy-middleware/-/http-proxy-middleware-1.0.0.tgz#4370a52766782e9c4f0be2ef79c3dd47aef5f428"
  integrity sha512-/s8lFX6rT43hSPqjjD8KNuu0SkPKY7uIdR6u9DCxVqCRhAvfKxGbVOixJsAT2mdpSnCyrGFAGoB39KFh6tmRxw==
  dependencies:
    http-proxy-middleware "*"
```

In `/app/node_modules/@types/http-proxy-middleware` is no `index.d.ts` . Just only `LICENSE`, `package.json` and `README.md`.

**Comment by G-Rath** [comment]
Your `@types/webpack-dev-server` is out of date - this version was published in `3.11.2` but you're using `3.11.0`.

I'm not having any issues with the types when testing this locally:

Would you be able to share a repo reproducing this issue?

I'm not having any issues with this:

```
import webpack = require('webpack');
import s = require('webpack-dev-server');
const compiler = webpack({});

console.log(s);

let server = new s(compiler, {
  publicPath2: '/assets/',
});
server.listen(8080);
```

```
{
  "name": "te",
  "version": "1.0.0",
  "main": "index.js",
  "license": "MIT",
  "devDependencies": {
    "@types/webpack": "^4.0.0",
    "@types/webpack-dev-server": "^3.11.2",
    "typescript": "^4.2.3",
    "webpack-dev-server": "^3.11.2"
  }
}
```

**Comment by razorness** [comment]
Thank you very much. You are right. :)
When I pin a `resolution` for `@types/webpack-dev-server` `^3.11.2`  in my `package.json`, everything is fine.
I faced this in a quasar project and `v3.11.0` is fixed by `@quasar/app`  `v2.2.1`.

**Pull request merged**